pylibs/unittests/test_real.py

```diff
@@ -0,0 +1,345 @@
+import socket
+import struct
+import threading
+import time
+import unittest
+from typing import Dict, List
+
+import grpc
+
+from OTNSTestCase import OTNSTestCase
+from otns.cli import errors, OTNS
+from otns.proto import visualize_grpc_pb2
+from otns.proto import visualize_grpc_pb2_grpc
+
+
+class UDPSignaler(object):
+    """Signaler for UDP messages.
+    """
+    def __init__(self, id: int):
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        self.source_addr = ("localhost", 9000 + id)
+        self.dest_addr = ("localhost", 9000)
+        self.sock.bind(self.source_addr)
+
+    def send_message(self, message: str) -> None:
+        data = message.encode("ascii")
+        event_packet = struct.pack("<QBH", 0, 5, len(data)) + data
+        self.sock.sendto(event_packet, self.dest_addr)
+
+    def close(self) -> None:
+        self.sock.close()
+
+
+class GRPCThread(threading.Thread):
+    """Thread to listen for expected gRPC stream response.
+    """
+    def __init__(self,
+                 stream,
+                 contain: List[str], 
+                 not_contain: List[str],
+                 exception_queue: queue.Queue,
+                 time_limit=2):
+        threading.Thread.__init__(self)
+        self.stream = stream
+        self.contain = contain
+        self.not_contain = not_contain
+        self.exception_queue = exception_queue
+        self.time_limit = time_limit
+
+    def run(self):
+        start_time = time.time()
+        try:
+            for event in self.stream:
+                if time.time() - start_time > self.time_limit:
+                    self.exception_queue.put(errors.ExpectationError(
+                            "Expectation not fulfilled within time limit"))
+                if (all(s in str(event) for s in self.contain)
+                    and not any(s in str(event) for s in self.not_contain)):
+                    return
+        except grpc._channel._MultiThreadedRendezvous as error:
+            self.exception_queue.put(
+                    errors.ExpectationError("gRPC service terminated"))
+        except grpc._channel._InactiveRpcError as error:
+            self.exception_queue.put(
+                    errors.ExpectationError("gRPC service terminated"))
+
+
+class GRPCClient(object):
+    """Listener for gRPC Visualize stream.
+    """
+    def __init__(self):
+        channel = grpc.insecure_channel("localhost:8999")
+        self.stub = visualize_grpc_pb2_grpc.VisualizeGrpcServiceStub(channel)
+        self.vis_stream = self.stub.Visualize(
+                visualize_grpc_pb2.VisualizeRequest())
+
+    def send_command(self, command: str):
+        self.stub.Command(
+        visualize_grpc_pb2.CommandRequest(command=command))
+
+    def expect_response(self,
+                        contain: List[str]=[],
+                        not_contain: List[str]=[]) -> GRPCThread:
+        exception_queue = queue.Queue()
+        expect_thread = GRPCThread(
```

Comment:
Should we cleanup the current stream (i.e. skip all events that are already received) before expecting new events?

Reply:
I'm not entirely sure how to properly implement clearing a stream buffer with Python gRPC APIs, so I ended up creating a new stream with every expect call.

Reply:
Using a new stream each time is risky and less efficient. Whenever a new stream is connected, OTNS sends many initial visualization events which could fail the checks here. 

So, whether or not it's a new stream, cleaning up the stream seems necessary before sending the command. 

Still, I would suggest using one stream for efficiency.  I am not sure if the stream can be read with timeouts or unblocked. If can't, we can create one dedicated Thread for reading visualization events from the stream and put the events to a queue for the main Thread to fetch.

Reply:
Makes sense

Reply:
@simonlingoogle I wonder what's the best way to implement this. By iterating the response stream, the call is blocking i.e. if I use a for loop, the loop continues to fetch next every time a new signal arrives, until exception occurs which signals that the stream has been terminated. This almost imitates a reactive API like `onNext`.
If I try to implement a separate worker thread to fetch the stream content into the buffer, i.e. creating this for the `GRPCClient`:
```Python
def fetch_stream(stream, buffer):
    try:
        for event in stream:
            buffer.append(str(event))
    except Exception as error:
        return

fetch_thread = threading.Thread(target=fetch_stream,
                                args=(self.stream, self.buffer))
fetch_thread.daemon = True
fetch_thread.start()
```
I can clear the buffer before creating the `GRPCThread`  as you suggested, but at the time the buffer would be empty, thus the thread will immediately return. It loses the ability to utilize the reactive nature of the gRPC streaming API. To mitigate that I would be running a time out loop `while ime.time() - start_time < self.time_limit` in the thread and use something like `time.sleep(0.1)`. This seems like a uglier way to get what we want.
On the other hand, why would cleaning up the stream be absolutely necessary? Iterating through some extra events before getting to the expected message should have no effect on what the expect code is doing, right? Especially because in practice every single test method creates a new OTNS instance by calling `setUp`, and a new stream.

Reply:
If we creates a new OTNS instance every time, it would fine to have new stream of `Visualize` RPC each time.
Thus the code is fine to me.